### DISCUSSION TITLE
[QUESTION] - How to apply same parameter in multiple places in SQL in slick 2.1.0

### DISCUSSION THREAD

**Discussion by sujit-baniya** [slick/slick#2655] [Questions]
```sql
SELECT id, first_name, last_name
           FROM table1
           WHERE col1 = $workItemId AND col2_uid = $encId
         UNION
           SELECT id, first_name, last_name
           FROM table2
           WHERE col3 = $workItemId AND col4 = $encId
```
In following example, how would I add query parameters without repeating. Currently I'm doing something like this

```scala
			val q = Q.query[(Int, Int, Int, Int), (Int, String, String)](
				"""SELECT id, first_name, last_name
           FROM table1
           WHERE col1 = ? AND col2_uid = ?
         UNION
           SELECT id, first_name, last_name
           FROM table2
           WHERE col3 = ? AND col4 = ?""")
			q.apply((workId, encId, workId, encId).list

```
I don't think this is the correct way of doing. In case the query is quite long in case I've to use param repetitively.
Please suggest any solution.

**Comment by jkugiya** [comment]
Would this be helpful?
https://scala-slick.org/doc/2.1.0/sql.html?highlight=literal#string-interpolation

**Reply by sujit-baniya** [reply]
@jkugiya I tried to use string interpolation as mentioned but results is weird errors.

```scala
: 8 more arguments than can be applied to method sql
```
But I've been providing the params correctly inside SQL.

**Reply by jkugiya** [reply]
On what grounds did you judge it to be "correct"?

**Reply by jkugiya** [reply]
https://github.com/scala/bug/issues/11281

Taking that message alone as a clue, it still seems to be caused by the implementation, although there is a problem with the compiler error message not being appropriate.

**Reply by sujit-baniya** [reply]
By "correct", I meant
The same SQL query is working on Scala 2.11.7 with slick 2.1.0 but it's not working on Scala 2.12 with 2.1.0

**Reply by jkugiya** [reply]
I can see that you have run some code. However, I do not know what code you are running. The code you exemplify above is not string interpolation, is it?

**Reply by sujit-baniya** [reply]
First I tried string interpolation, which gave argument-related issues.

So I tried with StaticQuery for which I've to provider same value multiple times for different placeholders

**Reply by jkugiya** [reply]
I believe the solution to the problem described at the beginning is string interpolation, but with the little information provided, it seems impossible to give further advice on the problem at compile time.

https://scala-slick.org/doc/2.1.0/sql.html?highlight=staticquery

Have you tried `+?` operator?

**Reply by sujit-baniya** [reply]
I've not tried it yet, but even though it's not safe, I will try it out...

**Reply by jkugiya** [reply]
If the behavior is as described in the documentation, there is no need to worry about SQL Injection due to combining strings since it is expanded into bind variables. 
What are the safety concerns?

> The SQL statement is the same for all calls: insert into coffees values (?,?,?,?,?)

**Reply by sujit-baniya** [reply]
Ah no!  I don't have safety concerns but I've to provide the same value in multiple places under the placeholder. It might be useful if I could define the param position below.

```scala
val q = Q.query[(Int, Int, Int, Int), (Int, String, String)](
				"""SELECT id, first_name, last_name
           FROM table1
           WHERE col1 = ?.1 AND col2_uid = ?.2
         UNION
           SELECT id, first_name, last_name
           FROM table2
           WHERE col3 = ?.1 AND col4 = ?.2""")
			q.apply((workId, encId).list
```

VS

```scala
val q = Q.query[(Int, Int, Int, Int), (Int, String, String)](
				"""SELECT id, first_name, last_name
           FROM table1
           WHERE col1 = ? AND col2_uid = ?
         UNION
           SELECT id, first_name, last_name
           FROM table2
           WHERE col3 = ? AND col4 = ?""")
			q.apply((workId, encId, workId, encId).list
```

Check the `?.1` and `?.2` in the first example is expected

**Reply by nafg** [reply]
Does this work? Note the nested tuple -- `((Int, Int), (Int, Int))` rather than `(Int, Int, Int, Int)`.

```scala
val tuple = (workId, encId)
val q = Q.query[((Int, Int), (Int, Int)), (Int, String, String)](
				"""SELECT id, first_name, last_name
           FROM table1
           WHERE col1 = ?.1 AND col2_uid = ?.2
         UNION
           SELECT id, first_name, last_name
           FROM table2
           WHERE col3 = ?.1 AND col4 = ?.2""")
			q.apply(tuple, tuple).list
```

If not, please provide a github repo that I can clone, run, and edit.

**Reply by jkugiya** [reply]
>  It might be useful if I could define the param position below.

As far as I know, slick doesn't have named or numbered placeholders, but it does have string interpolation or `+?` operator solves the problem of duplicate variables. (although `+?` is not in the current version.)

**Reply by sujit-baniya** [reply]
I've one example of string interpolation in slick 2.1.0 which is giving error

```scala
  def chargeQuery(workItemId: Int,
                  date: YmdDate,
                  proc: String,
                  mod1: Option[String],
                  mod2: Option[String],
                  chargeCodeSelect: Option[String],
                  chargeType: ChargeType,
                  patient_status_uid: Option[Int]): Q[Unit,ChargeCode] = {
    val m1 = mod1.orZero
    val m2 = mod2.orZero
    sql"""
      SELECT charge_master_uid, client_internal_code, client_proc_desc, charge_amt, client_mnemonic, code
      FROM (
          SELECT work_item_uid, charge_type, MAX(effective_date) AS effective_date, patient_status_uid
          FROM tbl_charge_master
          WHERE work_item_uid = $workItemId
                AND charge_type = ${chargeType.chargeKey}
                AND effective_date <= ${date.toString}
        ) AS EffectiveDate
        JOIN tbl_charge_master USING (work_item_uid,charge_type,effective_date)
        LEFT JOIN tbl_work_item_patient_status ON tbl_charge_master.patient_status_uid = tbl_work_item_patient_status.patient_status_uid
          AND EffectiveDate.work_item_uid = tbl_work_item_patient_status.work_item_uid
        WHERE cpt_hcpcs_code IN (${proc+m1+m2}, ${proc+m2+m1}, ${proc+m1}, ${proc+m2}, $proc)
          AND client_internal_code <> ''
        ORDER BY LENGTH(cpt_hcpcs_code) DESC,
                 client_internal_code = ${chargeCodeSelect.orZero} DESC
        LIMIT 1
       """
      .as[ChargeCode]
  }
```

Error:
```scala
ChargeHelper.scala:66:48: 8 more arguments than can be applied to method sql: (param: (Int, String, String, String, String, String, String, String, String))(implicit pconv: scala.slick.jdbc.SetParameter[(Int, String, String, String, String, String, String, String, String)])scala.slick.jdbc.SQLInterpolationResult[(Int, String, String, String, String, String, String, String, String)]
[error]                 AND charge_type = ${chargeType.chargeKey}

```

**Comment by nafg** [comment]
Please share a reproducer

**Reply by sujit-baniya** [reply]
@nafg The repo: https://github.com/sujit-baniya/slick-app

**Reply by sujit-baniya** [reply]
JDK Version: OpenJDK 17.0.6
OS: Ubuntu 22.04
